### PR TITLE
Add self-hosted Gitlab migration support

### PR DESCRIPTION
**Pull request by aqtrans**
This extends [PR#9084](https://github.com/go-gitea/gitea/pull/9084) to allow migrating from self-hosted Gitlab instances. 

Detection is based on adding a '#gitlab' fragment to the end of the given Clone Address.

### CODE REVIEW

**Comment by maicWorkGithub** [comment]
would you add an option `ignore certificate`?

**Comment by 6543** [comment]
no stale the main pull is not ready jet ...

**Comment by 6543** [comment]
@aqtrans can you rebase this pull?
since #9084 got merged?

**Comment by 6543** [comment]
you may want to wait for  #11136 to be merged :)

**Comment by 6543** [comment]
@aqtrans rebased for you. since I looked throu gitlab client's changes ... the lib update is only needed for reviews (#11137)

do you have a own gitlab instance to test with?

File: modules/migrations/gitlab.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ type GitlabDownloaderFactory struct {
 }
 
 // Match returns ture if the migration remote URL matched this downloader factory
+//   To allow self-hosting Gitlab instances, this matches based on the Host or a '#gitlab' fragment
```
Comment (6543) [inline modules/migrations/gitlab.go]:
```suggestion
// To allow self-hosting Gitlab instances, this matches based on the Host or a '#gitlab' fragment
```

File: modules/migrations/gitlab.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ type GitlabDownloaderFactory struct {
...
```
Comment (6543) [inline modules/migrations/gitlab.go]:
```suggestion
// To allow self-hosting Gitlab instances, this matches based on the Host or a '#gitlab' fragment
```

Comment (6543) [reply modules/migrations/gitlab.go]: again ... rebase deleted it somehow

**Review by 6543 — APPROVED** [review]
found a way to test it - worked just fine 🎉

**Comment by mrsdizzie** [comment]
I think adding #gitlab to the URL for detection feels like a confusing interface -- we should just add a drop down menu or something even if gitlab is the only option for now. I know there should be gitea/gogs at some point too (there is a PR for it) and maybe/hopefully more in the future.

**Review by 6543 — CHANGES_REQUESTED** [review]
x

**Comment by 6543** [comment]
@aqtrans are you on this?

**Comment by 6543** [comment]
@techknowlogick can we add this to v1.12 since it works only UI has change ...

**Comment by guillep2k** [comment]
I agree with @mrsdizzie here. There should be either a drop-down or a list of radio buttons (auto-detect, GitHub, GitLab, ... others to come). Slight problem here is that GitHub doesn't allow for a custom host address, which would be interesting for consistency (but it can't be helped since the api is provided by GitHub itself). Either way, IMHO it should not be a fragment of the URL. 😁

**Comment by techknowlogick** [comment]
another +1 for what @mrsdizzie said. In GitLab they have a migration page where you first select the forge type (gitlab, github, gitea, etc..) and then it will have different settings based on what you've selected (for example github we may want to have different help text about personal tokens needed for migration), and we could also split mirror and migration as it has lead to confusion in past.

@guillep2k our GitHub migration logic would also in theory support github enterprise too at a custom URL.

I am thankful for this PR as it has brought up these discussions. If it goes in I wouldn't be upset, and so I won't be a blocker, but I think the high-level needs to be reviewed first. Maybe it goes in temporarily until we can get improve the migration experience?

File: modules/migrations/gitlab.go [hunk 3]
```diff
@@ -43,6 +44,9 @@ func (f *GitlabDownloaderFactory) Match(opts base.MigrateOptions) (bool, error)
 	if strings.EqualFold(u.Host, "gitlab.com") && opts.AuthUsername != "" {
 		matched = true
 	}
+	if strings.EqualFold(u.Fragment, "gitlab") && opts.AuthUsername != "" {
```
Comment (lunny) [inline modules/migrations/gitlab.go:47]: For self-hosted gitlab, we need a new option on the UI to let user select.

**Comment by lunny** [comment]
A new wizard UI is better. I propose we have a migration type selection page before current.
There are four migration types you can chose:

- Mirror from a git service
- Migrate from a github service
- Migrate from a gitlab service
- Migrate from other git service

**Comment by 6543** [comment]
no stale just nobody wants to contribute the frontend 😅

**Comment by capi** [comment]
It would be great if this could be merged with the #gitlab fragment. I agree it's not "good UI" but at least it would work and I'm eagerly waiting to migrate some projects from my self-hosted Gitlab to Gitea. Thanks in advance!

**Comment by 6543** [comment]
@aqtrans do you have time, I wont for in the next weeks ...

**Pull request closed without merging**